File: map/map.py
from __future__ import annotations
from map.cell import Cell, TerrainTypes, gen_cell
from core.coord import Coord
from dataclasses import dataclass
from core.error import CoordinateAlreadyExistsError, CoordinateNotFoundError
from random import sample, randint
from organism.identity import Id
from core.basestorage import BaseStorage
from noise import pnoise2
from enum import Enum

class ScaleGenValues(Enum):
    LONG_LONG = 0.05
    LONG = 0.15
    MEDIUM = 0.30
    FRAGMENTED = 0.60

class Territory(BaseStorage[Coord, Cell]):
    def __init__(self, size_x:int, size_y:int):
        super().__init__()
        self.size_x = size_x
        self.size_y = size_y


    def _already_exists_error(self, key: Coord) -> None:
        raise CoordinateAlreadyExistsError('Coord {} already exists'.format(key))
    def _not_found_error(self, key: Coord) -> None:
        raise CoordinateNotFoundError('Coord {} was not found'.format(key))
    

class EntityMap(BaseStorage[Coord, Id]):
    def _already_exists_error(self, key: Coord) -> None:
        raise CoordinateAlreadyExistsError('Coord {} already exists'.format(key))
    def _not_found_error(self, key: Coord) -> None:
        raise CoordinateNotFoundError('Coord {} was not found'.format(key))
@dataclass(frozen=True)
class LigateCellView:
    coord:Coord
    cell:Cell

@dataclass(frozen=True)
class BlockData:
    cell:Cell
    id_creature: Id | None



class TerrainQuery:
    @staticmethod
    def all_unucoppied_coords(entity_map:EntityMap, territory:Territory) -> set[Coord]:
        return {c for c in territory.keys if not entity_map.exists(c)}
    @staticmethod
    def random_free_coord(territory:Territory, entity_map:EntityMap, n:int) -> list[Coord]:
        return sample([c for c in territory.keys if not entity_map.exists(c)], n)
    
    @staticmethod
    def get_coord_by_id(id:Id, entity_map:EntityMap) -> Coord:
        return entity_map.get_key_by_value(id) # <- O(n)
    
    



class TerrainMotor:
    @staticmethod
    def add_entity(id:Id, coord:Coord, territory:Territory, entity_map:EntityMap) -> None:
        if not territory.exists(coord):
            raise CoordinateNotFoundError('Coord {} was not found'.format(coord))
        entity_map.add(coord, id)

    @staticmethod
    def delete_entity_by_id(id:Id, entity_map:EntityMap) -> None:
        coord = entity_map.get_key_by_value(id) # <- O(n)
        entity_map.delete(coord)
        # O(n)


    @staticmethod
    def delete_coord(coord:Coord, territory:Territory, entity_map:EntityMap) -> None:
        territory.delete(coord)
        if entity_map.exists(coord):
            entity_map.delete(coord)
        
    @staticmethod
    def move(old_coord:Coord, new_coord:Coord, entity_map:EntityMap, territory:Territory) -> None:
        if not territory.exists(new_coord):
            raise CoordinateNotFoundError('Coord {} was not found in territory'.format(new_coord))
        id = entity_map.get(old_coord)
        entity_map.delete(old_coord)
        entity_map.add(new_coord, id)




class Geometry:
    @staticmethod
    def neighbors_x_y(coord:Coord, territory:Territory, entity_map:EntityMap, radius:Coord, include_self:bool) -> dict[Coord, BlockData]:
        neighbors_dict:dict[Coord, BlockData] = {}
        x = radius.x
        y = radius.y


        for row in range(-y, y+1):
            for column in range(-x, x+1):
                if row == 0 and column == 0:
                    if include_self:
                        neighbors_dict[coord] = BlockData(territory.get(coord), None)
                    continue


                coord_moved = Coord(y=row + coord.y, x=column + coord.x)

                if territory.exists(coord_moved):
                    creature_id = entity_map.require(coord_moved)
                    cell = territory.get(coord_moved)
                    data = BlockData(cell, creature_id)

                    neighbors_dict[coord_moved] = data
        
        return neighbors_dict

class TerrainFactory:
    @staticmethod
    def gen_seed() -> int:
        return randint(0, 10000)
    
    @staticmethod
    def value_to_cell_type(value:float) -> TerrainTypes:
        if value <= -0.3:
            return TerrainTypes.WATER
        elif value <= -0.1:
            return TerrainTypes.SAND
        elif value <= 0.7:
            return TerrainTypes.DIRT
        else:
            return TerrainTypes.ROCK
    
    @staticmethod
    def gen_terrain(radius:Coord, scale:float, seed:int) -> Territory:
        territory = Territory(radius.x, radius.y)

        for x in range(radius.x):
            for y in range(radius.y):
                value:float = pnoise2(x * scale, y * scale, base=seed) # type: ignore
                assert isinstance(value, float)
                terrain_type = TerrainFactory.value_to_cell_type(value)
                cell = gen_cell(terrain_type)
                territory.add(Coord(x, y), cell)
        return territory
    

File: src/karkinolution/terrain/map.cpp
#include <FastNoiseLite.h>
#include <karkinolution/terrain/map.hpp>
#include <karkinolution/utils/k_random.hpp>
#include <format>
#include <karkinolution/core/error.hpp>
#include <karkinolution/terrain/soil.hpp>
#include <algorithm>

namespace TerrainQuery {

    std::vector<Vec2> all_unoccupied_coords(const EntityMap& entity_map, const Territory& territory) {
        std::vector<Vec2> coords{};

        for (const auto& c : territory.keys()) {
            if (!entity_map.exists(c)) {
                coords.push_back(c);
            }
        }
        return coords;
    }

    std::vector<Vec2> random_unoccupied_coords(const EntityMap& entity_map, const Territory& territory, int n) {
        std::vector<Vec2> coords = TerrainQuery::all_unoccupied_coords(entity_map, territory);

        if (n <= 0 || coords.empty()) {
            return {};
        }

        const size_t requested = static_cast<size_t>(std::min(n, static_cast<int>(coords.size())));
        std::shuffle(coords.begin(), coords.end(), gen);
        coords.erase(coords.begin() + requested, coords.end());
        return coords;
    }

} 

// TerrainMotor
namespace TerrainMotor {

    void add_entity(Id id, Vec2 position, const Territory& territory, EntityMap& entity_map) {
        if (!territory.exists(position)) {
            throw CoordinateNotFoundError(std::format(
    "Failed to execute terrain operation: coordinate ({}, {}) was not found in the territory.", 
    position.x, position.y
));
        }
        entity_map.add(position, id);
    }

    void delete_coord(Vec2 coord, Territory&territory, EntityMap& entity_map) {
        if (!territory.exists(coord)) {
            throw CoordinateNotFoundError(std::format("Coordinate ({}, {}) was not found", coord.x, coord.y));
        }
        territory.del(coord);
        entity_map.del(coord);
    }

    void move(Vec2 old_coord, Vec2 new_coord, EntityMap& entity_map, Territory& territory) {
        if (!territory.exists(new_coord)) {
            throw CoordinateNotFoundError(std::format("Coordinate ({}, {}) was not found", new_coord.x, new_coord.y));
        }
        Id id = entity_map.at(old_coord);
        entity_map.del(old_coord);
        entity_map.add(new_coord, id);
    }
}
namespace Geometry {
    std::unordered_map<Vec2, BlockData> neighbors_x_y(Vec2 position, const Territory& territory, const EntityMap& entity_map, Vec2 radius, bool include_self) {
        std::unordered_map<Vec2, BlockData> neighbors{};
        int x_r = radius.x;
        int y_r = radius.y;
        
        for (int row = -y_r; row < y_r + 1; row++) {
            for (int column = -x_r; column < x_r + 1; column++) {
                if (row == 0 && column == 0 && include_self) {
                    const SoilPiece* piece = &territory.at(position);
                    neighbors[position] = BlockData(piece, std::nullopt); 
                    continue;
                }

                Vec2 coord_moved = Vec2(column + position.x, row + position.y);

                if (territory.exists(coord_moved)) {
                    const Id* id = entity_map.try_at(coord_moved);
                    const SoilPiece* piece = &territory.at(coord_moved);
                    std::optional<Id> creature_id;

                    if (id) {
                        creature_id = *id;
                    }
                    neighbors[coord_moved] = BlockData{piece, creature_id};
                }
            }
        }
        return neighbors;
    }   

}
// 
namespace TerrainFactory {

    int gen_seed() {
        return std::uniform_int_distribution<int>(1, 1000)(gen);
    }


    SoilTypes value_to_cell_type(float value) {
        if (value <= -0.3f) {
            return SoilTypes::WATER;
        }
        else if (value <= -0.1f) {
            return SoilTypes::SAND;
        }
        else if (value <= 0.7f) {
            return SoilTypes::DIRT;
        }
        else {
            return SoilTypes::ROCK;
        }
    }
    Territory gen_terrain(Vec2 radius, float scale, int seed) {
        Territory territory{static_cast<uint16_t>(radius.x), static_cast<uint16_t>(radius.y)};
        FastNoiseLite noise;
        noise.SetNoiseType(FastNoiseLite::NoiseType_Perlin);
        noise.SetSeed(seed);                                

        for (int x = 0; x < radius.x; ++x) {
            for (int y = 0; y < radius.y; ++y) {
                
                float value = noise.GetNoise(x * scale, y * scale);
                
                SoilTypes terrain_type = TerrainFactory::value_to_cell_type(value);
                
                auto cell = SoilF::gen_soil_piece(terrain_type);
                territory.add(Vec2{x, y}, std::move(cell));
            }
        }
        return territory; 
    }
}